src/graph.rs

```diff
@@ -0,0 +1,343 @@
+        GraphNodeRef(index)
+    }
+
+    /// Displays the contents of this graph.
+    pub fn display_with<'a>(&'a self, ctx: &'a Context) -> impl Display + 'a {
+        struct DisplayGraph<'a, 'tree>(&'a Graph<'tree>, &'a Context);
+
+        impl<'a, 'tree> Display for DisplayGraph<'a, 'tree> {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                let graph = self.0;
+                let ctx = self.1;
+                for (node_index, node) in graph.graph_nodes.iter().enumerate() {
+                    write!(
+                        f,
+                        "node {}\n{}",
+                        node_index,
+                        node.attributes.display_with(ctx, graph)
+                    )?;
+                    for (sink, edge) in &node.outgoing_edges {
+                        write!(
+                            f,
+                            "edge {} -> {}\n{}",
+                            node_index,
+                            *sink,
+                            edge.attributes.display_with(ctx, graph)
+                        )?;
+                    }
+                }
+                Ok(())
+            }
+        }
+
+        DisplayGraph(self, ctx)
+    }
+}
+
+impl<'tree> Index<SyntaxNodeRef> for Graph<'tree> {
+    type Output = Node<'tree>;
+    fn index(&self, index: SyntaxNodeRef) -> &Node<'tree> {
+        &self.syntax_nodes[&index]
+    }
+}
+
+impl Index<GraphNodeRef> for Graph<'_> {
+    type Output = GraphNode;
+    fn index(&self, index: GraphNodeRef) -> &GraphNode {
+        &self.graph_nodes[index.0 as usize]
+    }
+}
+
+impl<'tree> IndexMut<GraphNodeRef> for Graph<'_> {
+    fn index_mut(&mut self, index: GraphNodeRef) -> &mut GraphNode {
+        &mut self.graph_nodes[index.0 as usize]
+    }
+}
+
+/// A node in a graph
+pub struct GraphNode {
+    outgoing_edges: SmallVec<[(GraphNodeID, Edge); 8]>,
+    /// The set of attributes associated with this graph node
+    pub attributes: Attributes,
+}
+
+impl GraphNode {
+    fn new() -> GraphNode {
+        GraphNode {
+            outgoing_edges: SmallVec::new(),
+            attributes: Attributes::new(),
+        }
+    }
+
+    /// Adds an edge to this node.  There can be at most one edge connecting any two graph nodes;
+    /// the result indicates whether the edge is new (`Ok`) or already existed (`Err`).  In either
+    /// case, you also get a mutable reference to the [`Edge`][] instance for the edge.
+    pub fn add_edge(&mut self, sink: GraphNodeRef) -> Result<&mut Edge, &mut Edge> {
+        let sink = sink.0;
+        match self
+            .outgoing_edges
+            .binary_search_by_key(&sink, |(sink, _)| *sink)
+        {
+            Ok(index) => Err(&mut self.outgoing_edges[index].1),
+            Err(index) => {
+                self.outgoing_edges.insert(index, (sink, Edge::new()));
+                Ok(&mut self.outgoing_edges[index].1)
+            }
+        }
```

Comment:
Could we use `get_edge` or `get_edge_mut` (below) to reduce the duplication in the search process?

Reply:
We can't as written because of how the `Err` branches are handled differently.  I could pull the `binary_search_by_key` invocation out into a separate method...let me see how that looks.  I mention below that this amount of copy/paste doesn't bother me, but that one particular refactoring might be the right balance between our views.